src/mem/mem_scheduler.cc

```diff
@@ -0,0 +1,634 @@
+            DPRINTF(MemScheduler, "handleRequest: Blocking write request\n");
+            queue->sendWriteRetry = true;
+        }
+        if (pkt->isRead() || (unifiedQueue && pkt->isWrite())){
+            DPRINTF(MemScheduler, "handleRequest: Blocking read request\n");
+            queue->sendReadRetry = true;
+        }
+        return false;
+    }
+    if (pkt->isRead()){
+        stats.readReqs++;
+        entryTimes[pkt] = curTick();
+        DPRINTF(MemScheduler, "handleRequest: Pushing read request to readQueues[%d], size = %d\n", portId, queue->readQueue.size());
+        queue->push(pkt);
+        DPRINTF(MemScheduler, "handleRequest: Pushed read request to readQueues[%d], size = %d\n", portId, queue->readQueue.size());
+    } else{
+        stats.writeReqs++;
+        entryTimes[pkt] = curTick();
+        DPRINTF(MemScheduler, "handleRequest: Pushing write request to writeQueues[%d], size = %d\n", portId, queue->writeQueue.size());
+        queue->push(pkt);
+        DPRINTF(MemScheduler, "handleRequest: Pushed write request to writeQueues[%d], size = %d\n", portId, queue->writeQueue.size());
+    }
+    if (!nextReqEvent.scheduled()){
+        DPRINTF(MemScheduler, "handleRequest: Scheduling nextReqEvent in handleRequest\n");
+        schedule(nextReqEvent, curTick() + 100);
+    }
+    return true;
+}
+
+void
+MemScheduler::processNextReqEvent(){
+
+    uint64_t minCheck = -1;
+    RequestQueue *queue;
+    DPRINTF(MemScheduler, "processNextReqEvent: Finding the least recently visited non-empty queue\n");
+    for (auto &it : requestQueues){
+        if ((it.timesChecked < minCheck) && (!it.emptyRead() || it.serviceWrite())){
+            minCheck = it.timesChecked;
+            queue = &it;
+        }
+    }
+
+    stats.totalArbitrations++;
+    DPRINTF(MemScheduler, "processNextReqEvent: Least recently visited queue found, cpuPortId = %d, (timesChecked = %d) == (minChecked = %d), readQueue.size = %d, writeQueue.size = %d, emptyRead = %d, serviceWrite = %d\n", queue->cpuPortId, queue->timesChecked, minCheck, queue->readQueue.size(), queue->writeQueue.size(), queue->emptyRead(), queue->serviceWrite());
+
+    PacketPtr pkt;
+    if (!queue->serviceWrite()){
+        DPRINTF(MemScheduler, "processNextReqEvent: Servicing read request now\n");
+        pkt = queue->readQueue.front();
+    }
+    else{
+        DPRINTF(MemScheduler, "processNextReqEvent: Servicing write request now\n");
+        pkt = queue->writeQueue.front();
+    }
+    AddrRange addr_range = pkt->getAddrRange();
+    DPRINTF(MemScheduler, "processNextReqEvent: pkt->addr_range: %s\n", addr_range.to_string());
+    PortID memPortId = memPortMap.contains(addr_range)->second;
+    DPRINTF(MemScheduler, "processNextReqEvent: Looked up outgoing routing table for MemSidePort PortID: %d\n", memPortId);
+    auto memPort = find_if(memPorts.begin(), memPorts.end(), [memPortId](MemSidePort &obj){return obj.portId() == memPortId;});
```

Comment:
It looks like memPorts should be a map instead of whatever it is. `find_if` is a strange thing to use here.